src/expr/src/scalar/mod.rs

```diff
@@ -678,80 +750,80 @@ impl MirScalarExpr {
 
     /* #region `undistribute_and` and helper functions */
 
+    /// AND/OR undistribution to apply at each `ScalarExpr`.
     /// Transforms (a && b) || (a && c) into a && (b || c)
-    fn undistribute_and(&mut self) {
-        self.visit_mut(&mut |x| x.undistribute_and_helper());
-    }
-
-    /// AND undistribution to apply at each `ScalarExpr`.
-    fn undistribute_and_helper(&mut self) {
-        if let MirScalarExpr::CallBinary {
-            expr1,
-            expr2,
-            func: BinaryFunc::Or,
-        } = self
-        {
-            let mut ands0 = Vec::new();
-            expr1.harvest_ands(&mut ands0);
-
-            let mut ands1 = Vec::new();
-            expr2.harvest_ands(&mut ands1);
-
-            let mut intersection = Vec::new();
-            for expr in ands0.into_iter() {
-                if ands1.contains(&expr) {
-                    intersection.push(expr);
+    /// Transforms (a || b) && (a || c) into a || (b && c)
+    fn undistribute_and_or(&mut self) {
+        if let MirScalarExpr::CallBinary { expr1, expr2, func } = self {
+            if *func == BinaryFunc::Or || *func == BinaryFunc::And {
+                // We are trying to undistribute AND when we see
+                // `(a && b) || (a && c)`. Otherwise, we are trying to
+                // undistribute OR`.
+                let undistribute_and = *func == BinaryFunc::Or;
+                let mut operands0 = Vec::new();
```

Comment:
Nit: I'd name these temporary vectors as `operands1` and `operands2`, since they contain the operands of `expr1` and `expr2` respectively. I know the previous code was already using 0 and 1 as the suffixes.